### ISSUE TITLE
Regression: PATCH behavior along with Header version matching

### ISSUE THREAD

**Issue by dblock** [ruby-grape/grape#1355]
```ruby
require 'spec_helper'

describe Grape::API::Helpers do
  module PatchHelpersSpec
    class Patch_Public < Grape::API
      format :json
      version 'public-v1', using: :header, vendor: 'grape'

      get do
        { ok: 'public' }
      end
    end

    module AuthMethods
      def authenticate!

      end
    end

    class Patch_Private < Grape::API
      format :json
      version 'private-v1', using: :header, vendor: 'grape'

      helpers AuthMethods

      before do
        authenticate!
      end

      get do
        { ok: 'private' }
      end
    end

    class Main < Grape::API
      mount Patch_Public
      mount Patch_Private
    end
  end

  def app
    PatchHelpersSpec::Main
  end

  context 'default' do
    it 'public' do
      get '/', {}, 'HTTP_ACCEPT' => 'application/vnd.grape-public-v1+json'
      expect(last_response.status).to eq 200
      expect(last_response.body).to eq({ ok: 'public' }.to_json)
    end

    it 'private' do
      get '/', {}, 'HTTP_ACCEPT' => 'application/vnd.grape-private-v1+json'
      expect(last_response.status).to eq 200
      expect(last_response.body).to eq({ ok: 'private' }.to_json)
    end

    it 'default' do
      get '/'
      expect(last_response.status).to eq 200
      expect(last_response.body).to eq({ ok: 'public' }.to_json)
    end
  end

  context 'patch' do
    it 'public' do
      patch '/', {}, 'HTTP_ACCEPT' => 'application/vnd.grape-public-v1+json'
      expect(last_response.status).to eq 405
    end

    it 'private' do
      patch '/', {}, 'HTTP_ACCEPT' => 'application/vnd.grape-private-v1+json'
      expect(last_response.status).to eq 405
    end

    it 'default' do
      patch '/'
      expect(last_response.status).to eq 405
    end
  end
end
```

This passes on 0.15.0.

In 0.16.1 this fails in interesting ways.

```
~/source/grape/grape/dblock (patch-method-helpers)$ rspec spec/grape/api/patch_method_helpers_spec.rb 

Grape::API::Helpers
  default
    public
    private
    default
  patch
    public (FAILED - 1)
    private
    default

Failures:

  1) Grape::API::Helpers patch public
     Failure/Error: expect(last_response.status).to eq 405
     
       expected: 405
            got: 406
     
       (compared using ==)
     # ./spec/grape/api/patch_method_helpers_spec.rb:68:in `block (3 levels) in <top (required)>'

Finished in 0.05966 seconds (files took 0.34104 seconds to load)
6 examples, 1 failure

Failed examples:

rspec ./spec/grape/api/patch_method_helpers_spec.rb:66 # Grape::API::Helpers patch public
```

And

```
~/source/grape/grape/dblock (patch-method-helpers)$ rspec spec/grape/api/patch_method_helpers_spec.rb:65
Run options: include {:locations=>{"./spec/grape/api/patch_method_helpers_spec.rb"=>[65]}}

Grape::API::Helpers
  patch
    public (FAILED - 1)
    private (FAILED - 2)
    default (FAILED - 3)

Failures:

  1) Grape::API::Helpers patch public
     Failure/Error: authenticate!
     
     NoMethodError:
       undefined method `authenticate!' for #<#<Class:0x007fe599f99c38>:0x007fe59997fed8>
     # ./spec/grape/api/patch_method_helpers_spec.rb:27:in `block in <class:Patch_Private>'
     # ./lib/grape/endpoint.rb:334:in `instance_eval'
     # ./lib/grape/endpoint.rb:334:in `block (2 levels) in run_filters'
     # ./lib/grape/endpoint.rb:334:in `each'
     # ./lib/grape/endpoint.rb:334:in `block in run_filters'
     # /Users/dblock/.rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/notifications.rb:166:in `instrument'
     # ./lib/grape/endpoint.rb:333:in `run_filters'
     # ./lib/grape/router.rb:127:in `block in method_not_allowed'
     # ./lib/grape/router.rb:126:in `instance_eval'
     # ./lib/grape/router.rb:126:in `method_not_allowed'
     # ./lib/grape/router.rb:87:in `transaction'
     # ./lib/grape/router.rb:57:in `identity'
     # ./lib/grape/router.rb:44:in `block in call'
     # ./lib/grape/router.rb:104:in `with_optimization'
     # ./lib/grape/router.rb:43:in `call'
     # ./lib/grape/api.rb:112:in `call'
     # ./lib/grape/api.rb:45:in `call!'
     # ./lib/grape/api.rb:40:in `call'
     # /Users/dblock/.rvm/gems/ruby-2.3.0/gems/rack-test-0.6.3/lib/rack/mock_session.rb:30:in `request'
     # /Users/dblock/.rvm/gems/ruby-2.3.0/gems/rack-test-0.6.3/lib/rack/test.rb:244:in `process_request'
     # /Users/dblock/.rvm/gems/ruby-2.3.0/gems/rack-test-0.6.3/lib/rack/test.rb:85:in `patch'
     # ./spec/grape/api/patch_method_helpers_spec.rb:67:in `block (3 levels) in <top (required)>'

  2) Grape::API::Helpers patch private
     Failure/Error: authenticate!
     
     NoMethodError:
       undefined method `authenticate!' for #<#<Class:0x007fe599f99c38>:0x007fe59d09a030>
     # ./spec/grape/api/patch_method_helpers_spec.rb:27:in `block in <class:Patch_Private>'
     # ./lib/grape/endpoint.rb:334:in `instance_eval'
     # ./lib/grape/endpoint.rb:334:in `block (2 levels) in run_filters'
     # ./lib/grape/endpoint.rb:334:in `each'
     # ./lib/grape/endpoint.rb:334:in `block in run_filters'
     # /Users/dblock/.rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/notifications.rb:166:in `instrument'
     # ./lib/grape/endpoint.rb:333:in `run_filters'
     # ./lib/grape/router.rb:127:in `block in method_not_allowed'
     # ./lib/grape/router.rb:126:in `instance_eval'
     # ./lib/grape/router.rb:126:in `method_not_allowed'
     # ./lib/grape/router.rb:87:in `transaction'
     # ./lib/grape/router.rb:57:in `identity'
     # ./lib/grape/router.rb:44:in `block in call'
     # ./lib/grape/router.rb:104:in `with_optimization'
     # ./lib/grape/router.rb:43:in `call'
     # ./lib/grape/api.rb:112:in `call'
     # ./lib/grape/api.rb:45:in `call!'
     # ./lib/grape/api.rb:40:in `call'
     # /Users/dblock/.rvm/gems/ruby-2.3.0/gems/rack-test-0.6.3/lib/rack/mock_session.rb:30:in `request'
     # /Users/dblock/.rvm/gems/ruby-2.3.0/gems/rack-test-0.6.3/lib/rack/test.rb:244:in `process_request'
     # /Users/dblock/.rvm/gems/ruby-2.3.0/gems/rack-test-0.6.3/lib/rack/test.rb:85:in `patch'
     # ./spec/grape/api/patch_method_helpers_spec.rb:72:in `block (3 levels) in <top (required)>'

  3) Grape::API::Helpers patch default
     Failure/Error: authenticate!
     
     NoMethodError:
       undefined method `authenticate!' for #<#<Class:0x007fe599f99c38>:0x007fe59d0a2eb0>
     # ./spec/grape/api/patch_method_helpers_spec.rb:27:in `block in <class:Patch_Private>'
     # ./lib/grape/endpoint.rb:334:in `instance_eval'
     # ./lib/grape/endpoint.rb:334:in `block (2 levels) in run_filters'
     # ./lib/grape/endpoint.rb:334:in `each'
     # ./lib/grape/endpoint.rb:334:in `block in run_filters'
     # /Users/dblock/.rvm/gems/ruby-2.3.0/gems/activesupport-4.2.6/lib/active_support/notifications.rb:166:in `instrument'
     # ./lib/grape/endpoint.rb:333:in `run_filters'
     # ./lib/grape/router.rb:127:in `block in method_not_allowed'
     # ./lib/grape/router.rb:126:in `instance_eval'
     # ./lib/grape/router.rb:126:in `method_not_allowed'
     # ./lib/grape/router.rb:87:in `transaction'
     # ./lib/grape/router.rb:57:in `identity'
     # ./lib/grape/router.rb:44:in `block in call'
     # ./lib/grape/router.rb:104:in `with_optimization'
     # ./lib/grape/router.rb:43:in `call'
     # ./lib/grape/api.rb:112:in `call'
     # ./lib/grape/api.rb:45:in `call!'
     # ./lib/grape/api.rb:40:in `call'
     # /Users/dblock/.rvm/gems/ruby-2.3.0/gems/rack-test-0.6.3/lib/rack/mock_session.rb:30:in `request'
     # /Users/dblock/.rvm/gems/ruby-2.3.0/gems/rack-test-0.6.3/lib/rack/test.rb:244:in `process_request'
     # /Users/dblock/.rvm/gems/ruby-2.3.0/gems/rack-test-0.6.3/lib/rack/test.rb:85:in `patch'
     # ./spec/grape/api/patch_method_helpers_spec.rb:77:in `block (3 levels) in <top (required)>'

Finished in 0.0118 seconds (files took 0.33298 seconds to load)
3 examples, 3 failures

Failed examples:

rspec ./spec/grape/api/patch_method_helpers_spec.rb:66 # Grape::API::Helpers patch public
rspec ./spec/grape/api/patch_method_helpers_spec.rb:71 # Grape::API::Helpers patch private
rspec ./spec/grape/api/patch_method_helpers_spec.rb:76 # Grape::API::Helpers patch default
```

**Comment by dblock** [comment]
I could use some help from @namusyaka here ;(

**Comment by dblock** [comment]
The undefined method problem hasn't been fixed yet it seems.

**Comment by namusyaka** [comment]
Closed via #1362
